File: stock.rb
require 'gtk2'

module Cpg
	class Stock
		STATE = :"cpg-state"
		LINK = :"cpg-link"
		SENSOR = :"cpg-sensor"
		CHAIN = :"cpg-chain"
		CHAIN_BROKEN = :"cpg-chain-broken"
	
		def self.factory
			@factory
		end
	
		def self.icon_path(s)
			File.join(File.dirname(__FILE__), 'icons', s)
		end

		Gtk::Stock.add(Stock::STATE, 'State')
		Gtk::Stock.add(Stock::LINK, 'Link')
		Gtk::Stock.add(Stock::SENSOR, 'Sensor')
		
		Gtk::Stock.add(Stock::CHAIN, 'Chain')
		Gtk::Stock.add(Stock::CHAIN_BROKEN, 'Chain Broken')
	
		@factory = Gtk::IconFactory.new
		@factory.add(Cpg::Stock::STATE.to_s, Gtk::IconSet.new(Gdk::Pixbuf.new(icon_path('state.png'))))
		@factory.add(Cpg::Stock::LINK.to_s, Gtk::IconSet.new(Gdk::Pixbuf.new(icon_path('link.png'))))
		@factory.add(Cpg::Stock::SENSOR.to_s, Gtk::IconSet.new(Gdk::Pixbuf.new(icon_path('sensor.png'))))
		
		@factory.add(Cpg::Stock::CHAIN.to_s, Gtk::IconSet.new(Gdk::Pixbuf.new(icon_path('chain.png'))))
		@factory.add(Cpg::Stock::CHAIN_BROKEN.to_s, Gtk::IconSet.new(Gdk::Pixbuf.new(icon_path('chain-broken.png'))))
		
		@factory.add_default
		
		def self.chain_button
			but = Gtk::ToggleButton.new
			but.relief = Gtk::RELIEF_NONE
			
			im1 = Gtk::Image.new(icon_path('chain-broken.png'))
			im2 = Gtk::Image.new(icon_path('chain.png'))
			
			but.active = false
			but.image = im1
			
			but.signal_connect('toggled') do |x| 
				but.image = x.active? ? im2 : im1
				
				yield but if block_given?
			end
			
			but
		end
		
		def self.small_button(stock, &block)
			but = Gtk::Button.new
			
			rc = Gtk::RcStyle.new
			rc.ythickness = 0
			rc.xthickness = 0
			but.modify_style(rc)
			
			but.image = Gtk::Image.new(stock, Gtk::IconSize::MENU)
			but.relief = Gtk::ReliefStyle::NONE
			
			if block
				but.signal_connect('clicked') { |b| block.call(b) }
			end
			
			but
		end
		
		def self.close_button(&block)
			small_button(Gtk::Stock::CLOSE, &block)
		end
	end
end

File: Stock.cs
// Stock.cs created with MonoDevelop
// User: jesse at 19:36 24-3-2009
//
// To change standard headers go to Edit->Preferences->Coding->Standard Headers
//

using System;

namespace Cpg.Studio
{
	class Stock
	{
		public static string State = "cpg-state";
		public static string Link = "cpg-link";
		public static string Sensor = "cpg-sensor";
		public static string Relay = "cpg-relay";
		public static string Chain = "cpg-chain";
		public static string ChainBroken = "cpg-chain-broken";
		
		static Gtk.IconSet MakeIcons(Components.Renderers.Renderer renderer)
		{
			Gtk.IconSet s = new Gtk.IconSet();
			Gtk.IconSource source;
			
			Gtk.IconSize[] sizes = new Gtk.IconSize[] {
				Gtk.IconSize.Button,
				Gtk.IconSize.Dialog,
				Gtk.IconSize.Dnd,
				Gtk.IconSize.LargeToolbar,
				Gtk.IconSize.Menu,
				Gtk.IconSize.SmallToolbar
			};
			
			for (int i = 0; i < sizes.Length; ++i)
			{
				source = new Gtk.IconSource();
				
				int width, height;
				Gtk.Icon.SizeLookup(sizes[i], out width, out height);
				
				source.Pixbuf = renderer.Icon(width);
				source.Size = sizes[i];
				source.StateWildcarded = true;
				source.DirectionWildcarded = true;
				source.SizeWildcarded = false;
				
				s.AddSource(source);
			}
			
			return s;
		}
		
		static Stock()
		{
			Gtk.StockManager.Add(new Gtk.StockItem[] {
				new Gtk.StockItem(Stock.State, "State", 0, 0, null),
				new Gtk.StockItem(Stock.Link, "Link", 0, 0, null),
				new Gtk.StockItem(Stock.Relay, "Relay", 0, 0, null),
				new Gtk.StockItem(Stock.Chain, "Chain", 0, 0, null),
				new Gtk.StockItem(Stock.ChainBroken, "Chain Broken", 0, 0, null)
			});
			
			Gtk.IconFactory factory = new Gtk.IconFactory();

			factory.Add(Stock.State, MakeIcons(new Components.Renderers.State()));
			factory.Add(Stock.Link, new Gtk.IconSet(Gdk.Pixbuf.LoadFromResource("link.png")));
			factory.Add(Stock.Relay, MakeIcons(new Components.Renderers.Relay()));
			factory.Add(Stock.Chain, new Gtk.IconSet(Gdk.Pixbuf.LoadFromResource("chain.png")));
			factory.Add(Stock.ChainBroken, new Gtk.IconSet(Gdk.Pixbuf.LoadFromResource("chain-broken.png")));
			
			factory.AddDefault();
		}
	}
}
